src/Disks/S3/DiskS3.cpp

```diff
@@ -902,12 +922,371 @@ void DiskS3::shutdown()
     client->DisableRequestProcessing();
 }
 
-std::optional<DiskS3::ObjectMetadata> DiskS3::createObjectMetadata(const String & path) const
+void DiskS3::createFileOperationObject(const String & operation_name, UInt64 revision, const DiskS3::ObjectMetadata & metadata)
 {
-    if (send_metadata)
-        return (DiskS3::ObjectMetadata){{"path", path}};
+    const String key = "operations/r" + revisionToString(revision) + "-" + operation_name;
+    WriteBufferFromS3 buffer(client, bucket, s3_root_path + key, min_upload_part_size, max_single_part_upload_size, metadata);
+    buffer.write('0');
+    buffer.finalize();
+}
+
+void DiskS3::startup()
+{
+    if (!send_metadata)
+        return;
+
+    LOG_INFO(&Poco::Logger::get("DiskS3"), "Starting up disk {}", name);
+
+    /// Find last revision.
```

Comment:
Because we can lose all metadata information on the local disk including the last revision in separate file.
We can rely only on the data stored in S3.

Reply:
What about to store latest revision in S3 object called `latest_revision`? Or searching revision in such way expected to be cheap?

Reply:
Storing `latest_revision` every time to the separate object will cause overhead on each file write.
Search revision at start-up is cheap, just ~64 (binary search) head object requests.